common/network-types/src/session/state.rs

```diff
@@ -566,27 +583,75 @@ impl<const C: usize> SessionState<C> {
+                .ok_or(NetworkTypeError::Other("rate limiter ack rate invalid".into()))?,
+        );
 
-        let num_evicted = self.frame_reassembler.evict()?;
-        trace!(session_id = self.session_id, "evicted {num_evicted} frames");
+        // Rate limiter for retransmissions by the sender
+        let sender_retransmit = governor::RateLimiter::direct(
+            Quota::with_period(self.cfg.rto_base_sender)
+                .ok_or(NetworkTypeError::Other("rate limiter rto sender invalid".into()))?,
+        );
 
-        if self.cfg.enabled_features.contains(&SessionFeature::AcknowledgeFrames) {
-            self.acknowledge_segments().await?;
-        }
-        if self
-            .cfg
-            .enabled_features
-            .contains(&SessionFeature::RequestIncompleteFrames)
-        {
-            self.request_missing_segments().await?;
-        }
-        if self.cfg.enabled_features.contains(&SessionFeature::RetransmitFrames) {
-            self.retransmit_unacknowledged_frames().await?;
+        // Rate limiter for retransmissions by the receiver
+        let receiver_retransmit = governor::RateLimiter::direct(
+            Quota::with_period(self.cfg.rto_base_receiver)
+                .ok_or(NetworkTypeError::Other("rate limiter rto receiver invalid".into()))?,
+        );
+
+        loop {
+            let mut evict_fut = eviction_limiter.until_ready().boxed().fuse();
+            let mut ack_fut = maybe_fused_future(
+                self.cfg.enabled_features.contains(&SessionFeature::AcknowledgeFrames),
+                ack_rate_limiter.until_ready(),
+            );
+            let mut r_snd_fut = maybe_fused_future(
+                self.cfg.enabled_features.contains(&SessionFeature::RetransmitFrames),
+                sender_retransmit.until_ready(),
+            );
+            let mut r_rcv_fut = maybe_fused_future(
+                self.cfg
+                    .enabled_features
+                    .contains(&SessionFeature::RequestIncompleteFrames),
+                receiver_retransmit.until_ready(),
+            );
+            let mut is_done = maybe_fused_future(self.segment_egress_send.is_closed(), futures::future::ready(()));
+
+            // Futures in select_biased! are ordered from the least often happening first
+            if let Err(e) = futures::select_biased! {
+                _ = is_done => {
+                    Err(NetworkTypeError::Other("session writer has been closed".into()))
+                },
+                _ = r_rcv_fut => {
+                    self.request_missing_segments().await
+                },
+                _ = r_snd_fut => {
+                    self.retransmit_unacknowledged_frames().await
+                },
+                _ = ack_fut => {
+                    self.acknowledge_segments().await
+                },
+                 _ = evict_fut => {
+                    self.frame_reassembler.evict().map_err(NetworkTypeError::from)
+                },
```

Comment:
Would maybe the `Merge` behavior be better here? The evict fut has in a real traffic almost non-existent chance of being triggered, right?

Reply:
The evict fut is is ready about 10 times more often than the others. Therefore, it is the last one in the `select_biased!` so that the others actually get a chance of happening once they become ready as well (they take precedence before evict fut).